src/mscorlib/shared/System/Single.cs

```diff
@@ -230,6 +230,13 @@ public unsafe override int GetHashCode()
                 // Ensure that 0 and -0 have the same hash code
                 return 0;
             }
+
+            if (IsNaN(f))
+            {
+                // Ensure that all NaNs have the same hash code
+                return 1;
```

Comment:
This should just be `f = float.NaN;`, then the subsequent code will do the right thing. We don't want to use 1, as that will cause definite conflicts between NaN and `1.40129846E-45`.